File: app/Sources/GunkApp/Analyze/CapabilityFingerprint.swift
import Foundation

struct CapabilityFingerprint: Equatable, Sendable {
  let filePath: String
  let importedDependencies: Set<String>
  let routes: [RouteSurface]
  let publicExports: [ExportRef]
  let envVars: Set<String>
  let configKeys: Set<String>
  let namingTokens: Set<String>
  let capabilityHints: Set<CapabilityHint>
}

struct ClusterCapabilityFingerprint: Equatable, Sendable {
  let filePaths: Set<String>
  let importedDependencies: Set<String>
  let routes: [RouteSurface]
  let publicExports: [ExportRef]
  let envVars: Set<String>
  let configKeys: Set<String>
  let namingTokens: Set<String>
  let capabilityHints: Set<CapabilityHint>

  var hasPublicSurface: Bool {
    !routes.isEmpty || !publicExports.isEmpty
  }
}

struct CapabilityFingerprintBuilder: Sendable {
  private let lexicon: CapabilityLexicon
  private let routeDetector: RouteDetector

  init(lexicon: CapabilityLexicon = .default, routeDetector: RouteDetector = RouteDetector()) {
    self.lexicon = lexicon
    self.routeDetector = routeDetector
  }

  func fingerprints(
    files: [FileSymbols],
    manifests: [DependencyManifest],
    contentsByPath: [String: String]
  ) -> [CapabilityFingerprint] {
    let declaredDependencies = Set(manifests.flatMap(\.dependencies))

    return files.map { file in
      let contents = contentsByPath[file.path] ?? ""
      let importedDependencies = dependenciesImported(by: file, declaredDependencies: declaredDependencies)
      let hints = Set(importedDependencies.compactMap { lexicon.hint(for: $0) })

      return CapabilityFingerprint(
        filePath: file.path,
        importedDependencies: importedDependencies,
        routes: routeDetector.detect(path: file.path, contents: contents),
        publicExports: file.exports,
        envVars: envVars(in: contents),
        configKeys: configKeys(in: contents),
        namingTokens: namingTokens(for: file),
        capabilityHints: hints
      )
    }
    .sorted { $0.filePath < $1.filePath }
  }

  func aggregate(_ fingerprints: [CapabilityFingerprint], filePaths: Set<String>) -> ClusterCapabilityFingerprint {
    let included = fingerprints.filter { filePaths.contains($0.filePath) }

    return ClusterCapabilityFingerprint(
      filePaths: filePaths,
      importedDependencies: Set(included.flatMap(\.importedDependencies)),
      routes: included.flatMap(\.routes),
      publicExports: included.flatMap(\.publicExports),
      envVars: Set(included.flatMap(\.envVars)),
      configKeys: Set(included.flatMap(\.configKeys)),
      namingTokens: Set(included.flatMap(\.namingTokens)),
      capabilityHints: Set(included.flatMap(\.capabilityHints))
    )
  }

  private func dependenciesImported(by file: FileSymbols, declaredDependencies: Set<String>) -> Set<String> {
    Set(file.imports.compactMap { importRef in
      declaredDependencies.first { dependency in
        importMatchesDependency(importRef.moduleSpecifier, dependency: dependency)
      }
    })
  }

  private func importMatchesDependency(_ specifier: String, dependency: String) -> Bool {
    let normalizedSpecifier = normalizePackageName(packageRoot(for: specifier))
    let normalizedDependency = normalizePackageName(dependency)

    return normalizedSpecifier == normalizedDependency
      || normalizedSpecifier.hasPrefix("\(normalizedDependency)/")
      || normalizedDependency.hasPrefix("\(normalizedSpecifier)/")
  }

  private func packageRoot(for specifier: String) -> String {
    guard specifier.hasPrefix("@") else {
      return specifier.split(separator: "/").first.map(String.init) ?? specifier
    }

    let parts = specifier.split(separator: "/").map(String.init)
    guard parts.count >= 2 else {
      return specifier
    }

    return "\(parts[0])/\(parts[1])"
  }

  private func normalizePackageName(_ name: String) -> String {
    name.lowercased()
      .replacingOccurrences(of: "_", with: "-")
      .trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private func envVars(in contents: String) -> Set<String> {
    let patterns = [
      #"process\.env\.([A-Za-z_][A-Za-z0-9_]*)"#,
      #"process\.env\[['"]([A-Za-z_][A-Za-z0-9_]*)['"]\]"#,
      #"os\.environ\[['"]([A-Za-z_][A-Za-z0-9_]*)['"]\]"#,
      #"getenv\(['"]([A-Za-z_][A-Za-z0-9_]*)['"]\)"#,
      #"Environment\.GetEnvironmentVariable\(['"]([A-Za-z_][A-Za-z0-9_]*)['"]\)"#
    ]

    return Set(patterns.flatMap { pattern in
      analysisMatches(in: contents, pattern: pattern).compactMap(\.first)
    })
  }

  private func configKeys(in contents: String) -> Set<String> {
    let patterns = [
      #"config\.get\(['"]([A-Za-z0-9_\-\.]+)['"]\)"#,
      #"getConfig\(['"]([A-Za-z0-9_\-\.]+)['"]\)"#,
      #"settings\.([A-Za-z_][A-Za-z0-9_]*)"#
    ]

    return Set(patterns.flatMap { pattern in
      analysisMatches(in: contents, pattern: pattern).compactMap(\.first)
    })
  }

  private func namingTokens(for file: FileSymbols) -> Set<String> {
    let pathTokens = file.path
      .replacingOccurrences(of: ".", with: "/")
      .split(separator: "/")
      .map(String.init)
    let symbolTokens = file.symbols.flatMap { splitIdentifier($0.name) }
    let exportTokens = file.exports.flatMap { splitIdentifier($0.name) }

    return Set((pathTokens + symbolTokens + exportTokens)
      .flatMap(splitIdentifier)
      .map { $0.lowercased() }
      .filter { $0.count > 2 })
  }

  private func splitIdentifier(_ value: String) -> [String] {
    value
      .replacingOccurrences(of: "-", with: "_")
      .split(separator: "_")
      .flatMap { part in
        part.splitBeforeUppercase()
      }
  }
}

private extension Substring {
  func splitBeforeUppercase() -> [String] {
    var words: [String] = []
    var current = ""

    for character in self {
      if character.isUppercase, !current.isEmpty {
        words.append(current)
        current = ""
      }

      current.append(character)
    }

    if !current.isEmpty {
      words.append(current)
    }

    return words
  }
}

File: engine/src/analyze/capabilityFingerprint.ts
import type { ExportRef, FileSymbols } from "../models.js";
import { analysisMatches } from "./analysisRegex.js";
import {
  type CapabilityHint,
  capabilityHintKey,
  CapabilityLexicon,
} from "./capabilityLexicon.js";
import { type DependencyManifest } from "./dependencyManifest.js";
import { RouteDetector, type RouteSurface } from "./routeDetector.js";

export interface CapabilityFingerprint {
  filePath: string;
  importedDependencies: string[];
  routes: RouteSurface[];
  publicExports: ExportRef[];
  envVars: string[];
  configKeys: string[];
  namingTokens: string[];
  capabilityHints: CapabilityHint[];
}

export interface ClusterCapabilityFingerprint {
  filePaths: string[];
  importedDependencies: string[];
  routes: RouteSurface[];
  publicExports: ExportRef[];
  envVars: string[];
  configKeys: string[];
  namingTokens: string[];
  capabilityHints: CapabilityHint[];
  hasPublicSurface: boolean;
}

function dedupeStrings(values: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      out.push(value);
    }
  }
  return out;
}

function dedupeHints(hints: CapabilityHint[]): CapabilityHint[] {
  const seen = new Set<string>();
  const out: CapabilityHint[] = [];
  for (const hint of hints) {
    const key = capabilityHintKey(hint);
    if (!seen.has(key)) {
      seen.add(key);
      out.push(hint);
    }
  }
  return out;
}

function isUppercase(char: string): boolean {
  return char !== char.toLowerCase() && char === char.toUpperCase();
}

function splitBeforeUppercase(value: string): string[] {
  const words: string[] = [];
  let current = "";

  for (const char of value) {
    if (isUppercase(char) && current.length > 0) {
      words.push(current);
      current = "";
    }
    current += char;
  }

  if (current.length > 0) {
    words.push(current);
  }

  return words;
}

function splitIdentifier(value: string): string[] {
  return value
    .replace(/-/g, "_")
    .split("_")
    .filter((part) => part.length > 0)
    .flatMap((part) => splitBeforeUppercase(part));
}

export class CapabilityFingerprintBuilder {
  private readonly lexicon: CapabilityLexicon;
  private readonly routeDetector: RouteDetector;

  constructor(lexicon: CapabilityLexicon = CapabilityLexicon.default, routeDetector = new RouteDetector()) {
    this.lexicon = lexicon;
    this.routeDetector = routeDetector;
  }

  fingerprints(
    files: FileSymbols[],
    manifests: DependencyManifest[],
    contentsByPath: Record<string, string>,
  ): CapabilityFingerprint[] {
    const declaredDependencies = dedupeStrings(
      manifests.flatMap((manifest) => manifest.dependencies),
    ).sort();

    return files
      .map((file) => {
        const contents = contentsByPath[file.path] ?? "";
        const importedDependencies = this.dependenciesImported(file, declaredDependencies);
        const hints = dedupeHints(
          importedDependencies
            .map((dependency) => this.lexicon.hint(dependency))
            .filter((hint): hint is CapabilityHint => hint !== null),
        );

        return {
          filePath: file.path,
          importedDependencies,
          routes: this.routeDetector.detect(file.path, contents),
          publicExports: file.exports,
          envVars: this.envVars(contents),
          configKeys: this.configKeys(contents),
          namingTokens: this.namingTokens(file),
          capabilityHints: hints,
        };
      })
      .sort((lhs, rhs) => (lhs.filePath < rhs.filePath ? -1 : lhs.filePath > rhs.filePath ? 1 : 0));
  }

  aggregate(
    fingerprints: CapabilityFingerprint[],
    filePaths: Set<string>,
  ): ClusterCapabilityFingerprint {
    const included = fingerprints.filter((fingerprint) => filePaths.has(fingerprint.filePath));

    const routes = included.flatMap((fingerprint) => fingerprint.routes);
    const publicExports = included.flatMap((fingerprint) => fingerprint.publicExports);

    return {
      filePaths: [...filePaths],
      importedDependencies: dedupeStrings(
        included.flatMap((fingerprint) => fingerprint.importedDependencies),
      ),
      routes,
      publicExports,
      envVars: dedupeStrings(included.flatMap((fingerprint) => fingerprint.envVars)),
      configKeys: dedupeStrings(included.flatMap((fingerprint) => fingerprint.configKeys)),
      namingTokens: dedupeStrings(included.flatMap((fingerprint) => fingerprint.namingTokens)),
      capabilityHints: dedupeHints(included.flatMap((fingerprint) => fingerprint.capabilityHints)),
      hasPublicSurface:
        routes.length > 0 ||
        publicExports.length > 0 ||
        included.some((fingerprint) => fingerprint.capabilityHints.length > 0),
    };
  }

  private dependenciesImported(
    file: FileSymbols,
    declaredDependencies: string[],
  ): string[] {
    const matched: string[] = [];
    for (const importRef of file.imports) {
      const dependency = declaredDependencies.find((candidate) =>
        this.importMatchesDependency(importRef.moduleSpecifier, candidate),
      );
      if (dependency !== undefined) {
        matched.push(dependency);
      }
    }
    return dedupeStrings(matched);
  }

  private importMatchesDependency(specifier: string, dependency: string): boolean {
    const normalizedSpecifier = this.normalizePackageName(this.packageRoot(specifier));
    const normalizedDependency = this.normalizePackageName(dependency);
    const normalizedGroup = this.normalizePackageName(dependency.split(":")[0] ?? "");

    return (
      normalizedSpecifier === normalizedDependency ||
      normalizedSpecifier.startsWith(`${normalizedDependency}/`) ||
      normalizedDependency.startsWith(`${normalizedSpecifier}/`) ||
      (normalizedGroup.length > 0 &&
        (normalizedSpecifier === normalizedGroup ||
          normalizedSpecifier.startsWith(`${normalizedGroup}.`)))
    );
  }

  private packageRoot(specifier: string): string {
    if (specifier.startsWith("package:")) {
      const body = specifier.slice("package:".length);
      const first = body.split("/")[0];
      return first.length > 0 ? first : specifier;
    }

    if (!specifier.startsWith("@")) {
      const first = specifier.split("/")[0];
      return first.length > 0 ? first : specifier;
    }

    const parts = specifier.split("/");
    if (parts.length < 2) {
      return specifier;
    }

    return `${parts[0]}/${parts[1]}`;
  }

  private normalizePackageName(name: string): string {
    return name.toLowerCase().replace(/_/g, "-").trim();
  }

  private envVars(contents: string): string[] {
    const patterns = [
      String.raw`process\.env\.([A-Za-z_][A-Za-z0-9_]*)`,
      String.raw`process\.env\[['"]([A-Za-z_][A-Za-z0-9_]*)['"]\]`,
      String.raw`os\.environ\[['"]([A-Za-z_][A-Za-z0-9_]*)['"]\]`,
      String.raw`getenv\(['"]([A-Za-z_][A-Za-z0-9_]*)['"]\)`,
      String.raw`Environment\.GetEnvironmentVariable\(['"]([A-Za-z_][A-Za-z0-9_]*)['"]\)`,
    ];

    return dedupeStrings(
      patterns.flatMap((pattern) =>
        analysisMatches(contents, pattern)
          .map((groups) => groups[0])
          .filter((value): value is string => value !== undefined),
      ),
    );
  }

  private configKeys(contents: string): string[] {
    const patterns = [
      String.raw`config\.get\(['"]([A-Za-z0-9_\-\.]+)['"]\)`,
      String.raw`getConfig\(['"]([A-Za-z0-9_\-\.]+)['"]\)`,
      String.raw`settings\.([A-Za-z_][A-Za-z0-9_]*)`,
    ];

    return dedupeStrings(
      patterns.flatMap((pattern) =>
        analysisMatches(contents, pattern)
          .map((groups) => groups[0])
          .filter((value): value is string => value !== undefined),
      ),
    );
  }

  private namingTokens(file: FileSymbols): string[] {
    const pathTokens = file.path
      .replace(/\./g, "/")
      .split("/")
      .filter((component) => component.length > 0);
    const symbolTokens = file.symbols.flatMap((symbol) => splitIdentifier(symbol.name));
    const exportTokens = file.exports.flatMap((ref) => splitIdentifier(ref.name));

    return dedupeStrings(
      [...pathTokens, ...symbolTokens, ...exportTokens]
        .flatMap((token) => splitIdentifier(token))
        .map((token) => token.toLowerCase())
        .filter((token) => token.length > 2),
    );
  }
}
